pallets/dapp-staking/src/lib.rs

```diff
@@ -2707,55 +2707,60 @@ pub mod pallet {
+                    "Update has finished, skipping any action."
+                );
+                return Err(consumed_weight);
+            }
+
+            let mut bonus_update_state = init_bonus_update_state.clone();
             let mut entries_updated = 0u32;
 
-            let mut iter = if let Some(last_key_pair) = maybe_cursor.clone() {
-                StakerInfo::<T>::iter_from(StakerInfo::<T>::hashed_key_for(
-                    last_key_pair.0,
-                    last_key_pair.1,
-                ))
-            } else {
-                // Enable maintenance mode on the first run
-                // Likely to be already set in the runtime upgrade migration
-                ActiveProtocolState::<T>::mutate(|state| {
-                    state.maintenance = true;
-                });
-                log::warn!("Maintenance mode enabled.");
-                consumed_weight.saturating_add(T::DbWeight::get().reads_writes(1, 2));
-                StakerInfo::<T>::iter()
-            };
+            let mut iter =
+                if let BonusUpdateStateFor::<T>::InProgress(last_key_pair) = &bonus_update_state {
+                    StakerInfo::<T>::iter_from(StakerInfo::<T>::hashed_key_for(
+                        last_key_pair.0.clone(),
+                        last_key_pair.1.clone(),
+                    ))
+                } else {
+                    StakerInfo::<T>::iter()
+                };
 
+            let weight_margin = Self::update_weight_margin();
             while weight_limit
                 .saturating_sub(consumed_weight)
-                .all_gte(Self::update_weight_margin())
+                .all_gte(weight_margin)
             {
                 match Self::update_bonus_step(&mut iter) {
                     Ok((weight, updated_cursor)) => {
                         consumed_weight.saturating_accrue(weight);
                         entries_updated.saturating_inc();
-                        new_cursor = Some(updated_cursor);
+                        bonus_update_state = BonusUpdateStateFor::<T>::InProgress(updated_cursor);
                     }
                     Err(weight) => {
                         consumed_weight.saturating_accrue(weight);
-                        new_cursor = None; // Mark migration as complete
+                        bonus_update_state = BonusUpdateStateFor::<T>::Finished; // Mark migration as complete
                         break;
                     }
                 }
             }
 
-            if let Some(c) = new_cursor {
-                ActiveBonusUpdateCursor::<T>::put(c); // Save latest progress
-                consumed_weight = T::DbWeight::get().writes(1);
-            } else if maybe_cursor.is_some() {
-                consumed_weight = T::DbWeight::get().reads_writes(1, 3);
-                ActiveBonusUpdateCursor::<T>::take(); // Clear storage if update is finished
+            if bonus_update_state != init_bonus_update_state {
```

Comment:
this should always be different, you can do some cleanup by removing `init_bonus_update_state`